request.go

```diff
@@ -0,0 +1,251 @@
+	request, err := readHTTPRequest(req)
+	if err != nil {
+		util.GetLogger(req.Context()).WithError(err).Print("Error parsing HTTP headers")
+		return nil, util.MessageResponse(400, "Bad Request")
+	}
+	request.fields.Destination = destination
+
+	// The request fields are already in the form required by the specification
+	// So we can just serialise the request fields using the default marshaller
+	toVerify, err := json.Marshal(request.fields)
+	if err != nil {
+		util.GetLogger(req.Context()).WithError(err).Print("Error parsing JSON")
+		return nil, util.MessageResponse(400, "Invalid JSON")
+	}
+
+	if request.Origin() == "" {
+		message := "Missing \"Authorization: X-Matrix ...\" HTTP header"
+		util.GetLogger(req.Context()).WithError(err).Print(message)
+		return nil, util.MessageResponse(401, message)
+	}
+
+	results, err := keys.VerifyJSONs([]VerifyJSONRequest{{
+		ServerName: request.Origin(),
+		AtTS:       Timestamp(now.UnixNano() / 1000000),
+		Message:    toVerify,
+	}})
+	if err != nil {
+		message := "Error authenticating request"
+		util.GetLogger(req.Context()).WithError(err).Print(message)
+		return nil, util.MessageResponse(500, message)
+	}
+	if results[0].Result != nil {
+		message := "Invalid request signature"
+		util.GetLogger(req.Context()).WithError(results[0].Result).Print(message)
+		return nil, util.MessageResponse(401, message)
+	}
+
+	return request, util.JSONResponse{Code: 200, JSON: struct{}{}}
+}
+
+// Returns an error if there was a problem reading the content of the request
+func readHTTPRequest(req *http.Request) (*MatrixRequest, error) {
+	var result MatrixRequest
+
+	result.fields.Method = req.Method
+	result.fields.RequestURI = req.URL.RequestURI()
+
+	content, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+	if len(content) != 0 {
+		if req.Header.Get("Content-Type") != "application/json" {
+			return nil, fmt.Errorf(
+				"gomatrixserverlib: The request must be \"application/json\" not %q",
+				req.Header.Get("Content-Type"),
+			)
+		}
+		result.fields.Content = rawJSON(content)
+	}
+
+	for _, authorization := range req.Header["Authorization"] {
+		parts := strings.SplitN(authorization, " ", 2)
+		if parts[0] != "X-Matrix" {
+			continue
+		}
+		origin, key, sig := parseAuthorizationXMatrix(parts)
+		if origin == "" || key == "" || sig == "" {
+			return nil, fmt.Errorf("gomatrixserverlib: invalid X-Matrix authorization header")
+		}
+		if result.fields.Origin != "" && result.fields.Origin != origin {
+			return nil, fmt.Errorf("gomatrixserverlib: different origins in X-Matrix authorization headers")
+		}
+		result.fields.Origin = origin
+		if result.fields.Signatures == nil {
+			result.fields.Signatures = map[string]map[string]string{origin: map[string]string{key: sig}}
+		} else {
+			result.fields.Signatures[origin][key] = sig
+		}
+	}
+
+	return &result, nil
+}
+
+func parseAuthorizationXMatrix(headerParts []string) (origin, key, sig string) {
+	if len(headerParts) != 2 {
```

Comment:
Checks feel weirdly placed. The one place where `parseAuthorizationXMatrix` is called we have already checked that `[0]` is `X-Matrix`, but we do the length check here and never use `[0]` again, only `[1]`.

I suggest just passing the header value into this function and remove the length check.